adapters/pubmatic/pubmatic.go

```diff
@@ -65,23 +86,40 @@ func (a *PubmaticAdapter) Call(ctx context.Context, req *pbs.PBSRequest, bidder
 		var params pubmaticParams
 		err := json.Unmarshal(unit.Params, &params)
 		if err != nil {
+			errState = append(errState, fmt.Sprintf("BidID:%s;Error:%s;param:%s", unit.BidID, INVALID_PARAMS, unit.Params))
 			logf(PrepareLogMessage(pbReq.ID, params.PublisherId, unit.Code, unit.BidID,
 				fmt.Sprintf("Ignored bid: invalid JSON  [%s] err [%s]", unit.Params, err.Error())))
 			continue
 		}
 
 		if params.PublisherId == "" {
+			errState = append(errState, fmt.Sprintf("BidID:%s;Error:%s;param:%s", unit.BidID, MISSING_PUBID, unit.Params))
 			logf(PrepareLogMessage(pbReq.ID, params.PublisherId, unit.Code, unit.BidID,
 				fmt.Sprintf("Ignored bid: Publisher Id missing")))
 			continue
 		}
 		pubId = params.PublisherId
+
 		if params.AdSlot == "" {
+			errState = append(errState, fmt.Sprintf("BidID:%s;Error:%s;param:%s", unit.BidID, MISSING_ADSLOT, unit.Params))
 			logf(PrepareLogMessage(pbReq.ID, params.PublisherId, unit.Code, unit.BidID,
 				fmt.Sprintf("Ignored bid: adSlot missing")))
 			continue
 		}
 
+		// Parse Wrapper Extension i.e. ProfileID and VersionID only once per request
+		if wrapExt == "" && len(params.WrapExt) != 0 {
```

Comment:
params.WrapExt is a request level parameter at the Pubmatic side, so we want it to be parsed only once per request, even when it is passed multiple times in the request. In the code, whenever params.WrapExt is non-empty, wrapExt gets initialized to a value in that iteration (at line 120). Further, onwards it won't be re-initialized since wrapExt would be non empty.